File: pixels-common/src/main/java/io/pixelsdb/pixels/common/physical/natives/DirectRandomAccessFile.java
/*
 * Copyright 2023 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */
package io.pixelsdb.pixels.common.physical.natives;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.nio.ByteBuffer;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * The random accessible file that can be opened with o_direct flag.
 *
 * @author hank
 * @create 2023-02-02
 */
public class DirectRandomAccessFile implements PixelsRandomAccessFile
{
    private File file;
    private int fd;
    private long offset;
    private long length;
    private final int blockSize;
    private boolean bufferValid;
    private DirectBuffer smallBuffer;
    // Issue #403: largeBuffers may be updated by asynchronous read request, hence it must be thread-safe.
    private ConcurrentLinkedQueue<DirectBuffer> largeBuffers = new ConcurrentLinkedQueue<>();

    public DirectRandomAccessFile(File file) throws IOException
    {
        this.file = file;
        this.fd = DirectIoLib.open(file.getPath(), true);
        this.offset = 0;
        this.length = this.file.length();
        this.blockSize = DirectIoLib.FsBlockSize;
        this.bufferValid = false;
        try
        {
            this.smallBuffer = DirectIoLib.allocateBuffer(DirectIoLib.FsBlockSize);
        } catch (IllegalAccessException | InvocationTargetException e)
        {
            throw new IOException("failed to allocate buffer", e);
        }
    }

    @Override
    public void close() throws IOException
    {
        this.smallBuffer.close();
        this.smallBuffer = null;
        for (DirectBuffer largeBuffer : this.largeBuffers)
        {
            largeBuffer.close();
        }
        this.largeBuffers.clear();
        this.largeBuffers = null;
        DirectIoLib.close(fd);
        this.fd = -1;
        this.offset = 0;
        this.length = 0;
        this.file = null;
    }

    @Override
    public void readFully(byte[] b) throws IOException
    {
        ByteBuffer buffer = readFully(b.length);
        buffer.get(b);
    }

    @Override
    public void readFully(byte[] b, int off, int len) throws IOException
    {
        ByteBuffer buffer = readFully(len);
        buffer.get(b, off, len);
    }

    @Override
    public ByteBuffer readFully(int len) throws IOException
    {
        try
        {
            DirectBuffer buffer = DirectIoLib.allocateBuffer(len);
            DirectIoLib.read(this.fd, this.offset, buffer, len);
            this.seek(this.offset + len);
            this.largeBuffers.add(buffer);
            return buffer.getBuffer();
        } catch (IllegalAccessException | InvocationTargetException e)
        {
            throw new IOException("failed to allocate buffer", e);
        }
    }

    @Override
    public ByteBuffer readFully(long off, int len) throws IOException
    {
        try
        {
            DirectBuffer buffer = DirectIoLib.allocateBuffer(len);
            DirectIoLib.read(this.fd, off, buffer, len);
            this.largeBuffers.add(buffer);
            return buffer.getBuffer();
        } catch (IllegalAccessException | InvocationTargetException e)
        {
            throw new IOException("failed to allocate buffer", e);
        }
    }

    @Override
    public int skipBytes(int n) throws IOException
    {
        if (n <= 0)
        {
            return 0;
        }
        long off = this.offset;
        long newOff = Math.min(off + n, length);
        seek(newOff);
        return (int) (newOff - off);
    }

    private void populateBuffer() throws IOException
    {
        DirectIoLib.read(this.fd, this.offset, this.smallBuffer, this.blockSize);
        this.bufferValid = true;
    }

    @Override
    public boolean readBoolean() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.hasRemaining())
        {
            this.populateBuffer();
        }
        this.offset++;
        return this.smallBuffer.get() != 0;
    }

    @Override
    public byte readByte() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.hasRemaining())
        {
            this.populateBuffer();
        }
        this.offset++;
        return this.smallBuffer.get();
    }

    @Override
    public int readUnsignedByte() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.hasRemaining())
        {
            this.populateBuffer();
        }
        this.offset++;
        return this.smallBuffer.get() & 0xff;
    }

    @Override
    public short readShort() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.remaining() < Short.BYTES)
        {
            this.populateBuffer();
        }
        this.offset += Short.BYTES;
        return this.smallBuffer.getShort();
    }

    @Override
    public int readUnsignedShort() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.remaining() < Short.BYTES)
        {
            this.populateBuffer();
        }
        this.offset += Short.BYTES;
        return this.smallBuffer.getShort() & 0xffff;
    }

    @Override
    public char readChar() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.remaining() < Character.BYTES)
        {
            this.populateBuffer();
        }
        this.offset += Character.BYTES;
        return this.smallBuffer.getChar();
    }

    @Override
    public int readInt() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.remaining() < Integer.BYTES)
        {
            this.populateBuffer();
        }
        this.offset += Integer.BYTES;
        return this.smallBuffer.getInt();
    }

    @Override
    public long readLong() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.remaining() < Long.BYTES)
        {
            this.populateBuffer();
        }
        this.offset += Long.BYTES;
        return this.smallBuffer.getLong();
    }

    @Override
    public float readFloat() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.remaining() < Float.BYTES)
        {
            this.populateBuffer();
        }
        this.offset += Float.BYTES;
        return this.smallBuffer.getFloat();
    }

    @Override
    public double readDouble() throws IOException
    {
        if (!this.bufferValid || this.smallBuffer.remaining() < Double.BYTES)
        {
            this.populateBuffer();
        }
        this.offset += Double.BYTES;
        return this.smallBuffer.getDouble();
    }

    @Override
    public void seek(long offset) throws IOException
    {
        if (this.bufferValid && offset > this.offset - this.smallBuffer.position() &&
                offset < this.offset + this.smallBuffer.remaining())
        {
            this.smallBuffer.forward((int) (offset - this.offset));
        }
        else
        {
            this.bufferValid = false;
        }
        this.offset = offset;
    }

    @Override
    public long length()
    {
        return this.length;
    }
}

File: cpp/pixels-common/lib/physical/natives/DirectRandomAccessFile.cpp
/*
 * Copyright 2023 PixelsDB.
 *
 * This file is part of Pixels.
 *
 * Pixels is free software: you can redistribute it and/or modify
 * it under the terms of the Affero GNU General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Pixels is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * Affero GNU General Public License for more details.
 *
 * You should have received a copy of the Affero GNU General Public
 * License along with Pixels.  If not, see
 * <https://www.gnu.org/licenses/>.
 */

/*
 * @author liyu
 * @create 2023-03-02
 */
#include "physical/natives/DirectRandomAccessFile.h"

#include "physical/natives/DirectIoLib.h"
#include "utils/ConfigFactory.h"

#include <cstdio>
#include <malloc.h>
#include "profiler/CountProfiler.h"
#include "profiler/TimeProfiler.h"
#include "physical/allocator/OrdinaryAllocator.h"
#include "physical/allocator/BufferPoolAllocator.h"

DirectRandomAccessFile::DirectRandomAccessFile(const std::string& file)
{
    FILE* fp = fopen(file.c_str(), "r");
    // checking if the file exist or not
    if (fp == nullptr)
    {
        throw std::runtime_error("DirectRandomAccessFile: File not found or fd exceeds the limitation. ");
    }
    fseek(fp, 0L, SEEK_END);
    // calculating the size of the file
    len = ftell(fp);
    // closing the file
    fclose(fp);
    fsBlockSize = std::stoi(ConfigFactory::Instance().getProperty("localfs.block.size"));
    enableDirect = ConfigFactory::Instance().boolCheckProperty("localfs.enable.direct.io");
    if (enableDirect)
    {
        fd = open(file.c_str(), O_RDONLY | O_DIRECT);
    }
    else
    {
        fd = open(file.c_str(), O_RDONLY);
        smallBuffer = std::make_shared<ByteBuffer>(fsBlockSize);
    }
    offset = 0;

    bufferValid = false;
    directIoLib = std::make_shared<DirectIoLib>(fsBlockSize);
    try
    {
        smallDirectBuffer = directIoLib->allocateDirectBuffer(fsBlockSize, true);
    }
    catch (...)
    {
        throw std::runtime_error("failed to allocate buffer");
    }
    allocator = std::make_shared<BufferPoolAllocator>();
}

void DirectRandomAccessFile::close()
{
    largeBuffers.clear();

    if (fd != -1 && ::close(fd) != 0)
    {
        throw std::runtime_error("File is not closed properly");
    }
    fd = -1;
    offset = 0;
    len = 0;
}

std::shared_ptr<ByteBuffer> DirectRandomAccessFile::readFully(int len)
{
    if (enableDirect)
    {
        auto directBuffer = directIoLib->allocateDirectBuffer(len, true);
        auto buffer = directIoLib->read(fd, offset, directBuffer, len);
        seek(offset + len);
        largeBuffers.emplace_back(directBuffer);
        return buffer;
    }
    else
    {
        auto buffer = allocator->allocate(len);
        if (pread(fd, buffer->getPointer(), len, offset) == -1)
        {
            throw std::runtime_error("pread fail");
        }
        seek(offset + len);
        largeBuffers.emplace_back(buffer);
        return buffer;
    }
}

std::shared_ptr<ByteBuffer> DirectRandomAccessFile::readFully(int len, std::shared_ptr<ByteBuffer> bb)
{
    if (enableDirect)
    {
        auto buffer = directIoLib->read(fd, offset, bb, len);
        seek(offset + len);
        return buffer;
    }
    else
    {
        if (pread(fd, bb->getPointer(), len, offset) == -1)
        {
            throw std::runtime_error("pread fail");
        }
        seek(offset + len);
        return std::make_shared<ByteBuffer>(*bb, 0, len);
    }
}


long DirectRandomAccessFile::length()
{
    return len;
}

void DirectRandomAccessFile::seek(long off)
{
    if (bufferValid && off > offset - smallBuffer->getReadPos() &&
        off < offset + smallBuffer->bytesRemaining())
    {
        smallBuffer->setReadPos(off - offset + smallBuffer->getReadPos());
    }
    else
    {
        bufferValid = false;
    }
    offset = off;
}

long DirectRandomAccessFile::readLong()
{
    if (!bufferValid || smallBuffer->bytesRemaining() < sizeof(long))
    {
        populatedBuffer();
    }
    offset += sizeof(long);
    return smallBuffer->getLong();
}

int DirectRandomAccessFile::readInt()
{
    if (!bufferValid || smallBuffer->bytesRemaining() < sizeof(int))
    {
        populatedBuffer();
    }
    offset += sizeof(int);
    return smallBuffer->getInt();
}

char DirectRandomAccessFile::readChar()
{
    if (!bufferValid || smallBuffer->bytesRemaining() < sizeof(char))
    {
        populatedBuffer();
    }
    offset += sizeof(char);
    return smallBuffer->getChar();
}

void DirectRandomAccessFile::populatedBuffer()
{
    if (enableDirect)
    {
        smallBuffer = directIoLib->read(fd, offset, smallDirectBuffer, fsBlockSize);
        bufferValid = true;
    }
    else
    {
        if (pread(fd, smallBuffer->getPointer(), fsBlockSize, offset) == -1)
        {
            throw std::runtime_error("pread fail");
        }
        smallBuffer->resetPosition();
        bufferValid = true;
    }
}
